auth/internal/transport/cba.go

```diff
@@ -0,0 +1,411 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transport
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/auth/internal/transport/cert"
+	"cloud.google.com/go/compute/metadata"
+	"github.com/google/s2a-go"
+	"github.com/google/s2a-go/fallback"
+	"google.golang.org/grpc/credentials"
+)
+
+const (
+	mTLSModeAlways = "always"
+	mTLSModeNever  = "never"
+	mTLSModeAuto   = "auto"
+
+	configEndpointSuffix = "googleAutoMtlsConfiguration"
+
+	// Experimental: if true, the code will try MTLS with S2A as the default for transport security. Default value is false.
+	googleAPIUseS2AEnv     = "EXPERIMENTAL_GOOGLE_API_USE_S2A"
+	googleAPIUseCertSource = "GOOGLE_API_USE_CLIENT_CERTIFICATE"
+	googleAPIUseMTLS       = "GOOGLE_API_USE_MTLS_ENDPOINT"
+	googleAPIUseMTLSOld    = "GOOGLE_API_USE_MTLS"
+)
+
+var (
+	// The period an MTLS config can be reused before needing refresh.
+	configExpiry = time.Hour
+
+	// mtlsEndpointEnabledForS2A checks if the endpoint is indeed MTLS-enabled, so that we can use S2A for MTLS connection.
+	mtlsEndpointEnabledForS2A = func() bool {
+		// TODO(xmenxk): determine this via discovery config.
+		return true
+	}
+
+	// mdsMTLSAutoConfigSource is an instance of reuseMTLSConfigSource, with metadataMTLSAutoConfig as its config source.
+	mtlsOnce                sync.Once
+	mdsMTLSAutoConfigSource mtlsConfigSource
+)
+
+// Options is a struct that is duplicated information from the individual
+// transport packages in order to avoid cyclic deps.
+type Options struct {
+	Endpoint            string
+	DefaultEndpoint     string
+	DefaultMTLSEndpoint string
+	ClientCertProvider  cert.Provider
+	Client              *http.Client
```

Comment:
Would it be possible to add docs to each field? Maybe copy the docs from the original declarations in the individual transport packages, and to each, add a note stating from where the field was duplicated?

Reply:
I can for sure clarify the struct in the main docs for this struct, but I am hesitant to copy field docs for this kind of struct. It just means there is one more place that documentation can drift. These all should correlate 1:1 with another struct already. But if you feel strongly here we can add fields docs.

Reply:
> I can for sure clarify the struct in the main docs for this struct

Yes, please add docs where appropriate, thanks.